### ISSUE TITLE
Document DIY Docker builds

### ISSUE THREAD

**Issue by jml** [commercialhaskell/stack#3129]
My use case is I want to build my Haskell projects with Stack inside a Docker container without having Stack installed. This is so my colleagues can just run `make` on a project without having to install dependencies beyond Docker. 

Essentially, I want to do what `stack --docker build` does, but I want to do it by hand.

Requirements:
* a build image that has stack installed and set up
* a way of caching artefacts on the host system between builds
* local, non-Docker `stack` to continue to work

You can see how I've cobbled this together at the moment:
* [project Makefile](https://github.com/weaveworks-experiments/compare-revisions/blob/master/Makefile)
* [build image definition](https://github.com/weaveworks/build-tools/pull/88/files)

The essential ingredients are:
* run `docker` very carefully:

    ```
	docker run --rm -ti \
		-e STACK_IN_CONTAINER=1 \
		-e STACK_ROOT=$(HOME) \
		-e HOME=$(shell pwd)/.stack-work/docker/_home \
		-v $(HOME):$(HOME) \
		-v $(HOME)/.stack:$(HOME)/.stack \
		-v $(shell pwd):$(shell pwd) \
		-v $(shell pwd)/.stack-work/docker/_home:$(shell pwd)/.stack-work/docker/_home \
		quay.io/weaveworks/build-haskell:latest $@
    ```
* run `stack` very carefully:

    ```
	stack --internal-re-exec-version=1.3.2 \
		--internal-docker-entrypoint "DockerEntrypoint {deUser = Nothing}" \
		--docker \
		--local-bin-path $(BINARY_DIR) \
		--copy-bins \
		build
    ```

`BINARY_DIR` is a Makefile constant which is used in another place in order to gather together the images.

I don't really understand what's going on here. In particular, I don't really know:
* what goes in each of these directories
* where the build artefacts (esp. for dependencies) are cached
* where the package index is stored
* why the whole of `$HOME` needs to be mounted
* what the effect of `STACK_IN_CONTAINER` is
* why the `--internal-docker-entrypoint` and `--internal-re-exec-version` flags are needed

And probably more things also.

**Comment by mgsloan** [comment]
Good point, it'd be good to have more docs detailing how such things work.  For now taking a look at the code is a good option

**Comment by jml** [comment]
That's more or less how I figured this out. Still don't know how to avoid rebuilding all my deps all the time. Pointers appreciated.

**Comment by AlistairB** [comment]
@jml I have something like this working - https://github.com/AlistairB/docker-stack-haskell

Deps are cached via a docker volume. It isn't quite perfect. I sort of wish I could have GHC installed within the Dockerfile but don't know how to separate this setup from the packages. Anyway, overall I quite like the way it is working.

**Comment by lwm** [comment]
@jml you could also try the mailing list https://groups.google.com/forum/#!forum/haskell-stack.
